File: backend/services/websiteCategorizer.js
const axios = require('axios');
const cheerio = require('cheerio');
const Website = require('../models/Website');

class WebsiteCategorizer {
    constructor() {
        this.industry_keywords = {
            'banking': ['bank', 'loan', 'mortgage', 'credit', 'debit', 'account', 'finance', 'investment'],
            'healthcare': ['hospital', 'clinic', 'doctor', 'medical', 'health', 'patient', 'treatment', 'medicine'],
            'education': ['school', 'university', 'college', 'course', 'student', 'teacher', 'education', 'learning'],
            'ecommerce': ['shop', 'store', 'product', 'cart', 'checkout', 'price', 'sale', 'buy'],
            'technology': ['software', 'hardware', 'tech', 'digital', 'computer', 'internet', 'app', 'mobile'],
            'government': ['gov', 'government', 'official', 'public', 'service', 'department', 'ministry'],
            'entertainment': ['movie', 'music', 'game', 'entertainment', 'show', 'media', 'stream'],
            'news': ['news', 'article', 'report', 'journal', 'press', 'media', 'headline'],
            'travel': ['travel', 'tourism', 'hotel', 'flight', 'booking', 'vacation', 'trip'],
            'real_estate': ['property', 'real estate', 'house', 'apartment', 'rent', 'sale', 'home']
        };

        this.website_types = {
            'corporate': ['about', 'company', 'corporate', 'business', 'enterprise'],
            'ecommerce': ['shop', 'store', 'cart', 'checkout', 'product'],
            'informational': ['about', 'info', 'information', 'guide', 'help'],
            'social': ['login', 'signup', 'profile', 'user', 'community', 'forum'],
            'service': ['service', 'support', 'help', 'contact', 'assistance'],
            'blog': ['blog', 'article', 'post', 'news', 'update'],
            'portfolio': ['portfolio', 'work', 'projects', 'gallery', 'showcase'],
            'directory': ['directory', 'listing', 'catalog', 'index', 'search']
        };
    }

    async fetchWebsiteContent(url) {
        try {
            const response = await axios.get(url, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
                }
            });
            return response.data;
        } catch (error) {
            throw new Error(`Failed to fetch website content: ${error.message}`);
        }
    }

    analyzeContent(htmlContent) {
        try {
            const $ = cheerio.load(htmlContent);
            
            // Extract text content
            const textContent = $('body').text().toLowerCase();
            
            // Extract meta tags
            const metaTags = {
                title: $('title').text().toLowerCase(),
                description: $('meta[name="description"]').attr('content')?.toLowerCase() || '',
                keywords: $('meta[name="keywords"]').attr('content')?.toLowerCase() || ''
            };
            
            // Analyze industry
            const industryScores = this.analyzeIndustry(textContent, metaTags);
            const primaryIndustry = Object.entries(industryScores)
                .reduce((a, b) => a[1] > b[1] ? a : b)[0];
            
            // Analyze website type
            const typeScores = this.analyzeWebsiteType(textContent, metaTags);
            const primaryType = Object.entries(typeScores)
                .reduce((a, b) => a[1] > b[1] ? a : b)[0];
            
            // Analyze functionality
            const functionality = this.analyzeFunctionality($);
            
            // Analyze target audience
            const audience = this.analyzeTargetAudience(textContent, metaTags);
            
            return {
                primary_industry: primaryIndustry,
                industry_confidence: industryScores[primaryIndustry],
                website_type: primaryType,
                type_confidence: typeScores[primaryType],
                functionality,
                target_audience: audience,
                meta_information: metaTags
            };
        } catch (error) {
            throw new Error(`Error analyzing website content: ${error.message}`);
        }
    }

    analyzeIndustry(text, metaTags) {
        const scores = {};
        const combinedText = `${text} ${metaTags.title} ${metaTags.description} ${metaTags.keywords}`;
        
        for (const [industry, keywords] of Object.entries(this.industry_keywords)) {
            scores[industry] = keywords.reduce((score, keyword) => {
                const regex = new RegExp(`\\b${keyword}\\b`, 'gi');
                return score + (combinedText.match(regex) || []).length * 0.1;
            }, 0);
        }
        
        // Normalize scores
        const maxScore = Math.max(...Object.values(scores));
        if (maxScore > 0) {
            Object.keys(scores).forEach(key => {
                scores[key] = scores[key] / maxScore;
            });
        }
        
        return scores;
    }

    analyzeWebsiteType(text, metaTags) {
        const scores = {};
        const combinedText = `${text} ${metaTags.title} ${metaTags.description} ${metaTags.keywords}`;
        
        for (const [type, keywords] of Object.entries(this.website_types)) {
            scores[type] = keywords.reduce((score, keyword) => {
                const regex = new RegExp(`\\b${keyword}\\b`, 'gi');
                return score + (combinedText.match(regex) || []).length * 0.1;
            }, 0);
        }
        
        // Normalize scores
        const maxScore = Math.max(...Object.values(scores));
        if (maxScore > 0) {
            Object.keys(scores).forEach(key => {
                scores[key] = scores[key] / maxScore;
            });
        }
        
        return scores;
    }

    analyzeFunctionality($) {
        const functionality = [];
        
        if ($('form').length) functionality.push('forms');
        if ($('input[type="search"]').length) functionality.push('search');
        if ($('a[href*="login"], a[href*="signin"], a[href*="signup"], a[href*="register"]').length) {
            functionality.push('user_authentication');
        }
        if ($('a[href*="cart"], a[href*="checkout"], a[href*="basket"]').length) {
            functionality.push('ecommerce');
        }
        if ($('a[href*="blog"], a[href*="news"], a[href*="article"]').length) {
            functionality.push('content_management');
        }
        if ($('a[href*="social"], a[href*="share"], a[href*="facebook"], a[href*="twitter"]').length) {
            functionality.push('social_integration');
        }
        
        return functionality;
    }

    analyzeTargetAudience(text, metaTags) {
        const combinedText = `${text} ${metaTags.title} ${metaTags.description} ${metaTags.keywords}`;
        
        const b2bKeywords = ['business', 'enterprise', 'corporate', 'wholesale', 'b2b'];
        const b2cKeywords = ['consumer', 'customer', 'retail', 'personal', 'individual'];
        
        const b2bScore = b2bKeywords.reduce((score, keyword) => {
            const regex = new RegExp(`\\b${keyword}\\b`, 'gi');
            return score + (combinedText.match(regex) || []).length;
        }, 0);
        
        const b2cScore = b2cKeywords.reduce((score, keyword) => {
            const regex = new RegExp(`\\b${keyword}\\b`, 'gi');
            return score + (combinedText.match(regex) || []).length;
        }, 0);
        
        if (b2bScore > b2cScore) return 'B2B';
        if (b2cScore > b2bScore) return 'B2C';
        return 'General';
    }

    async processWebsite(url) {
        try {
            // Check if website already exists
            let website = await Website.findOne({ url });
            
            if (!website) {
                website = new Website({ url });
            }
            
            website.status = 'processing';
            await website.save();
            
            // Fetch and analyze website content
            const htmlContent = await this.fetchWebsiteContent(url);
            const categories = this.analyzeContent(htmlContent);
            
            // Update website with categories
            website.categories = categories;
            website.status = 'completed';
            website.processed_pages = 1; // For now, we're only processing the main page
            website.last_processed = new Date();
            
            await website.save();
            
            return website;
        } catch (error) {
            if (website) {
                website.status = 'failed';
                website.error_message = error.message;
                await website.save();
            }
            throw error;
        }
    }
}

module.exports = new WebsiteCategorizer(); 

File: app/website_categorizer.py
from bs4 import BeautifulSoup
import requests
import re
from typing import Dict, List, Tuple
import logging

logger = logging.getLogger(__name__)

class WebsiteCategorizer:
    def __init__(self):
        self.industry_keywords = {
            'banking': ['bank', 'loan', 'mortgage', 'credit', 'debit', 'account', 'finance', 'investment'],
            'healthcare': ['hospital', 'clinic', 'doctor', 'medical', 'health', 'patient', 'treatment', 'medicine'],
            'education': ['school', 'university', 'college', 'course', 'student', 'teacher', 'education', 'learning'],
            'ecommerce': ['shop', 'store', 'product', 'cart', 'checkout', 'price', 'sale', 'buy'],
            'technology': ['software', 'hardware', 'tech', 'digital', 'computer', 'internet', 'app', 'mobile'],
            'government': ['gov', 'government', 'official', 'public', 'service', 'department', 'ministry'],
            'entertainment': ['movie', 'music', 'game', 'entertainment', 'show', 'media', 'stream'],
            'news': ['news', 'article', 'report', 'journal', 'press', 'media', 'headline'],
            'travel': ['travel', 'tourism', 'hotel', 'flight', 'booking', 'vacation', 'trip'],
            'real_estate': ['property', 'real estate', 'house', 'apartment', 'rent', 'sale', 'home']
        }

        self.website_types = {
            'corporate': ['about', 'company', 'corporate', 'business', 'enterprise'],
            'ecommerce': ['shop', 'store', 'cart', 'checkout', 'product'],
            'informational': ['about', 'info', 'information', 'guide', 'help'],
            'social': ['login', 'signup', 'profile', 'user', 'community', 'forum'],
            'service': ['service', 'support', 'help', 'contact', 'assistance'],
            'blog': ['blog', 'article', 'post', 'news', 'update'],
            'portfolio': ['portfolio', 'work', 'projects', 'gallery', 'showcase'],
            'directory': ['directory', 'listing', 'catalog', 'index', 'search']
        }

    def analyze_content(self, html_content: str) -> Dict:
        """Analyze website content and return categorization results"""
        try:
            soup = BeautifulSoup(html_content, 'html.parser')
            
            # Extract text content
            text_content = soup.get_text().lower()
            
            # Extract meta tags
            meta_tags = {
                'title': soup.title.string.lower() if soup.title else '',
                'description': soup.find('meta', {'name': 'description'})['content'].lower() if soup.find('meta', {'name': 'description'}) else '',
                'keywords': soup.find('meta', {'name': 'keywords'})['content'].lower() if soup.find('meta', {'name': 'keywords'}) else ''
            }
            
            # Analyze industry
            industry_scores = self._analyze_industry(text_content, meta_tags)
            primary_industry = max(industry_scores.items(), key=lambda x: x[1])[0]
            
            # Analyze website type
            type_scores = self._analyze_website_type(text_content, meta_tags)
            primary_type = max(type_scores.items(), key=lambda x: x[1])[0]
            
            # Analyze functionality
            functionality = self._analyze_functionality(soup)
            
            # Analyze target audience
            audience = self._analyze_target_audience(text_content, meta_tags)
            
            return {
                'primary_industry': primary_industry,
                'industry_confidence': industry_scores[primary_industry],
                'website_type': primary_type,
                'type_confidence': type_scores[primary_type],
                'functionality': functionality,
                'target_audience': audience,
                'meta_information': meta_tags
            }
            
        except Exception as e:
            logger.error(f"Error analyzing website content: {str(e)}")
            return {
                'error': str(e),
                'primary_industry': 'unknown',
                'website_type': 'unknown'
            }

    def _analyze_industry(self, text: str, meta_tags: Dict) -> Dict[str, float]:
        """Analyze and score potential industries"""
        scores = {industry: 0.0 for industry in self.industry_keywords.keys()}
        
        # Combine all text for analysis
        combined_text = f"{text} {meta_tags['title']} {meta_tags['description']} {meta_tags['keywords']}"
        
        for industry, keywords in self.industry_keywords.items():
            for keyword in keywords:
                # Count keyword occurrences
                count = len(re.findall(r'\b' + keyword + r'\b', combined_text))
                scores[industry] += count * 0.1  # Weight factor for each occurrence
        
        # Normalize scores
        max_score = max(scores.values()) if scores.values() else 1
        if max_score > 0:
            scores = {k: v/max_score for k, v in scores.items()}
        
        return scores

    def _analyze_website_type(self, text: str, meta_tags: Dict) -> Dict[str, float]:
        """Analyze and score potential website types"""
        scores = {type_: 0.0 for type_ in self.website_types.keys()}
        
        # Combine all text for analysis
        combined_text = f"{text} {meta_tags['title']} {meta_tags['description']} {meta_tags['keywords']}"
        
        for type_, keywords in self.website_types.items():
            for keyword in keywords:
                # Count keyword occurrences
                count = len(re.findall(r'\b' + keyword + r'\b', combined_text))
                scores[type_] += count * 0.1  # Weight factor for each occurrence
        
        # Normalize scores
        max_score = max(scores.values()) if scores.values() else 1
        if max_score > 0:
            scores = {k: v/max_score for k, v in scores.items()}
        
        return scores

    def _analyze_functionality(self, soup: BeautifulSoup) -> List[str]:
        """Analyze website functionality based on HTML elements"""
        functionality = []
        
        # Check for common functionality indicators
        if soup.find('form'):
            functionality.append('forms')
        if soup.find('input', {'type': 'search'}):
            functionality.append('search')
        if soup.find('a', href=re.compile(r'login|signin|signup|register')):
            functionality.append('user_authentication')
        if soup.find('a', href=re.compile(r'cart|checkout|basket')):
            functionality.append('ecommerce')
        if soup.find('a', href=re.compile(r'blog|news|article')):
            functionality.append('content_management')
        if soup.find('a', href=re.compile(r'social|share|facebook|twitter')):
            functionality.append('social_integration')
        
        return functionality

    def _analyze_target_audience(self, text: str, meta_tags: Dict) -> str:
        """Analyze target audience based on content"""
        combined_text = f"{text} {meta_tags['title']} {meta_tags['description']} {meta_tags['keywords']}"
        
        # Simple keyword-based analysis
        b2b_keywords = ['business', 'enterprise', 'corporate', 'wholesale', 'b2b']
        b2c_keywords = ['consumer', 'customer', 'retail', 'personal', 'individual']
        
        b2b_score = sum(len(re.findall(r'\b' + keyword + r'\b', combined_text)) for keyword in b2b_keywords)
        b2c_score = sum(len(re.findall(r'\b' + keyword + r'\b', combined_text)) for keyword in b2c_keywords)
        
        if b2b_score > b2c_score:
            return 'B2B'
        elif b2c_score > b2b_score:
            return 'B2C'
        else:
            return 'General' 